sound/soc/sof/pm.c

```diff
@@ -422,3 +438,41 @@ int snd_sof_suspend(struct device *dev)
 	return sof_suspend(dev, false);
 }
 EXPORT_SYMBOL(snd_sof_suspend);
+
+int snd_sof_set_dsp_power_state(struct snd_sof_dev *sdev,
+				enum sof_d0_substate d0_substate)
+{
+	u32 flags;
+	int ret;
+
+	dev_dbg(sdev->dev, "setting ADSP D0 substate:%d\n", d0_substate);
+
+	/* Todo: destroy/reload unused pipelines/widgets in FW */
+
+	/* do platform specific set_state */
+	ret = snd_sof_dsp_set_power_state(sdev, d0_substate);
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"error: d0i3_exit ipc error %d\n", ret);
+		return ret;
+	}
+
+	if (d0_substate == SOF_DSP_D0I3)
+		flags = SOF_PM_PG_STREAMING;
+	else
+		flags = SOF_PM_PPG | SOF_PM_PCG;
```

Comment:
@keyonjie is this necessary? I am assuming if the requested state is not D0i3, it is D0, In that case, isnt it automatically assumed that power gating and clock gating should be disabled?

Reply:
from the PR(https://github.com/thesofproject/sof/pull/1729) that @mmaka1 sent, for D0i3 exit transform(D0i3->D0), we need to disable Power Gating, so need to set SOF_PM_PPG and SOF_PM_PCG bits of flags with the PM_GATE IPC, it is a little bit convoluted, isn't it? It uses double negation (*disable* *prevent* power gating, which means enable power gating) at D0i3 case.

Reply:
@keyonjie are you sure about this. The way that I understood was that when the D0i3 state is requested, the driver also has the option of requesting not to power_gate or clock_gate. When transitioning to D0, I think it should be a given that both power_gating and clock_gating is disabled.

Reply:
@ranj063 from the PM_GATE IPC defined in @mmaka1's PR, the FW doesn't know about D0 or D0i3(with the same PM_GATE IPC), we are using those bits in the flag to let the FW know about whether we are asking for power saving or not, the here it is actually the mapping from D0Ix to flag bits.

Reply:
And for the mapping itself, yes, we can make it configurable in the driver side if it is needed in the future.

Reply:
@keyonjie I see. So its actually the other way around then? You would automatically enable pg/cg when entering D0i3?

Reply:
> @keyonjie I see. So its actually the other way around then? You would automatically enable pg/cg when entering D0i3?

yes, will set PPG and PCG bits to be 0 at D0i3, that actually enables pg/cg.

Reply:
thanks @keyonjie. makes sense then